source/common/event/file_event_impl.h

```diff
@@ -41,5 +45,136 @@ class FileEventImpl : public FileEvent, ImplBase {
+   * @param enabled_events supplied the event of activate().
+   */
+  virtual void onEventActivated(uint32_t activated_events) PURE;
+};
+
+// Return the enabled events except EV_CLOSED. This implementation is generally good since only
+// epoll supports EV_CLOSED but the entire envoy code base supports another poller. The event owner
+// must assume EV_CLOSED is never activated. Also event owner must tolerate that OS could notify
+// events which are not actually triggered.
+// TODO(lambdai): Add support of delivering EV_CLOSED.
+class DefaultEventListener : public EventListener {
+public:
+  ~DefaultEventListener() override = default;
+
+  // Return both read and write.
+  uint32_t triggeredEvents() override { return pending_events_ & (~Event::FileReadyType::Closed); }
+
+  void onEventEnabled(uint32_t enabled_events) override { pending_events_ = enabled_events; }
+
+  void onEventActivated(uint32_t activated_events) override {
+    ephermal_events_ |= activated_events;
+  }
+
+  uint32_t getAndClearEphemeralEvents() override { return std::exchange(ephermal_events_, 0); }
+
+private:
+  // The persisted interested events and ready events.
+  uint32_t pending_events_{};
+  // The events set by activate() and will be cleared after the io callback.
+  uint32_t ephermal_events_{};
+};
+
+// A FileEvent implementation which is used to drive BufferedIoSocketHandle.
+class UserSpaceFileEventImpl : public FileEvent, Logger::Loggable<Logger::Id::io> {
+public:
+  ~UserSpaceFileEventImpl() override {
+    if (schedulable_.enabled()) {
+      schedulable_.cancel();
+    }
+  }
+
+  // Event::FileEvent
+  void activate(uint32_t events) override {
+    // Only supported event types are set.
+    ASSERT((events & (FileReadyType::Read | FileReadyType::Write | FileReadyType::Closed)) ==
+           events);
+    event_listener_.onEventActivated(events);
+    if (!schedulable_.enabled()) {
+      schedulable_.scheduleCallbackNextIteration();
+    }
+  }
+
+  void setEnabled(uint32_t events) override {
+    // Only supported event types are set.
+    ASSERT((events & (FileReadyType::Read | FileReadyType::Write | FileReadyType::Closed)) ==
+           events);
+    event_listener_.onEventEnabled(events);
+    bool was_enabled = schedulable_.enabled();
+    if (!was_enabled) {
+      schedulable_.scheduleCallbackNextIteration();
+    }
+    ENVOY_LOG(trace, "User space file event {} set events {}. Will {} reschedule.",
+              static_cast<void*>(this), events, was_enabled ? "not " : "");
+  }
+
+  EventListener& getEventListener() { return event_listener_; }
+  void onEvents() { cb_(); }
+  friend class UserSpaceFileEventFactory;
+  friend class Network::BufferedIoSocketHandleImpl;
+
+private:
+  UserSpaceFileEventImpl(Event::FileReadyCb cb, uint32_t events,
+                         SchedulableCallback& schedulable_cb)
+      : schedulable_(schedulable_cb), cb_([this, cb]() {
+          auto all_events = getEventListener().triggeredEvents();
```

Comment:
Saying that the userspace fd is readable and writable after every call to setEnabled and several other transitions would "work", but you'ld end up with a lot of spurious extra events which would result in an efficiency hit.

We can probably do better by having a read buffer, a write buffer and a half closed flags as part of the user-space event listener.  Behaviors to implement (partial list):
- A call to setEnable(read) would schedule a read event if the read buffer is non-empty.
- A call to setEnable(write) would schedule a write event if the write buffer is not full.
- Bytes added to the read buffer would schedule a read event if the buffer was previously empty.
- Bytes removed from the write buffer would schedule a write event if the buffer was previously full.
- When read buffer is marked half closed, schedule a read event if buffer is empty.

Reply:
Thank you for the enumerations! 

I agree the goal is to lessen the miss wake up. 

The current implementation could raise as many as 2x the events, since readable could trigger writable and vice versa.

My question is can we add these in follow up PRs? 

Each step from 2x to 1x events is error prone and might need extra state flag and test coverage.

I feel it is smooth if we start with a correct but less efficient implementation.

Reply:
You could argue that the Close event work around is a correctness issue.  I guess the question for you is: "How do we ensure this code is not used until finished and tested?"

Reply:
There are massive of ways to misuse the underlying components like this BufferedIoSocketHandle and UserSpaceFileEvent. We probably cannot prevent another piece of code construct these WIP code.

However, my expected flow is to construct by dispatcher. See the full picture PR here: https://github.com/envoyproxy/envoy/pull/13361/files#diff-c39ba1f2f907b54929fdc875227dad65c198262bfe1d48dbe51b9118c42d16d9R62-R69

Along with the half close: I have basic test cases in this full picture as well: https://github.com/envoyproxy/envoy/pull/13361/files#diff-d50ad3bb92cf82447311aabe7dded76c69de7fee4e65a8a020f186dfc6c924c0R107-R135

That full PR need sync with this PR regularly, and that is not covering as many complicated cases as OS socket, but still it has come coverage on setEnable, flowcontrol and half close

Reply:
> We can probably do better by having a read buffer, a write buffer and a half closed flags as part of the user-space event listener. Behaviors to implement (partial list):

Yes, it was several iterations back before I introduce EvenetListener. At that time all the logic are mingled, buffer data consumption and production are mutating the states everywhere.

Then I realize it's not an error to trigger events when I read the epoll implementation.

My plan is to add an EfficientEventListener, watching more than enable() and active() but also buffer data and state flags in IoHandle. That implementation should deliver less events than the current DefaultEventListener at the cost of error prone.

That's why I feel a dumb but easier DefaultEventListener could smooth the future `EfficientEventListener`(not well-named yet)

Reply:
Add a decent unit test cases on UserSpaceFileEventImpl